### PR TITLE
Capped Drucker-Prager plasticity added

### LINKED ISSUE #8474: Add ComputeCappedDruckerPragerStress to TensorMechanics
### Description of the enhancement or error report

This is Drucker-Prager plasticity, capped with tensile and compressive failure envelopes.  This will derive from PQPlasticModel

### Rationale for the enhancement or information for reproducing the error

Rock mechanics, especially where tensile failure causes complete loss of compressive strength

### PR DESCRIPTION
**Pull request by WilkAndy**
Small modifications to PQPlasticModel and TensorMechanicsPlasticDruckerPrager

Fixes #8474

### CODE REVIEW

File: modules/tensor_mechanics/include/materials/ComputeCappedDruckerPragerStress.h [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+   * _stress_return_type is set to no_compression, and every time
+   * the compressive yield function is evaluated it is set -(very large).
+   * If the compressive yield function yf[2] >= 0 at the trial stress
+   * then clearly the quadpoint is not going to fail in tension, so
+   * _stress_return_type is set to no_tension, and every time the
+   * tensile yield function is evaluated it is set -(very large).
+   * Otherwise (and at the very end after return-map) _stress_return_type
+   * is set to nothing_special.
+   */
+  enum class StressReturnType {
```
Comment (permcody) [inline modules/tensor_mechanics/include/materials/ComputeCappedDruckerPragerStress.h]: 👍

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_inner_tip.i [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform2_inner_tip
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_inner_tip.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_lode_zero.i [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    value = 0
+  [../]
+  [./dp]
+    type = TensorMechanicsPlasticDruckerPrager
+    mc_cohesion = mc_coh
+    mc_friction_angle = mc_phi
+    mc_dilation_angle = mc_psi
+    mc_interpolation_scheme = lode_zero
+    internal_constraint_tolerance = 1 # irrelevant here
+    yield_function_tolerance = 1      # irrelevant here
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_lode_zero.i]: Missing closing block here.

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_lode_zero.i [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform2_lode_zero
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_lode_zero.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_native.i [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    value = 0
+  [../]
+  [./dp]
+    type = TensorMechanicsPlasticDruckerPrager
+    mc_cohesion = mc_coh
+    mc_friction_angle = mc_phi
+    mc_dilation_angle = mc_psi
+    mc_interpolation_scheme = native
+    internal_constraint_tolerance = 1 # irrelevant here
+    yield_function_tolerance = 1      # irrelevant here
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_native.i]: Missing closing block here.

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_native.i [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform2_native
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_native.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_outer_tip.i [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    value = 0
+  [../]
+  [./dp]
+    type = TensorMechanicsPlasticDruckerPrager
+    mc_cohesion = mc_coh
+    mc_friction_angle = mc_phi
+    mc_dilation_angle = mc_psi
+    mc_interpolation_scheme = outer_tip
+    internal_constraint_tolerance = 1 # irrelevant here
+    yield_function_tolerance = 1      # irrelevant here
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_outer_tip.i]: Missing closing block here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_outer_tip.i [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform2_outer_tip
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_outer_tip.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_inner_edge.i [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform3_inner_edge
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_inner_edge.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_inner_tip.i [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform3_inner_tip
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_inner_tip.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_lode_zero.i [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform3_lode_zero
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_lode_zero.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_native.i [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform3_native
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_native.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_outer_tip.i [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform3_outer_tip
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform3_outer_tip.i]: Spacing here

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_inner_edge.i [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+  type = Transient
+[]
+
+
+[Outputs]
+  file_base = small_deform2_inner_edge
+  exodus = false
+  [./csv]
+    type = CSV
+    [../]
```
Comment (permcody) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_inner_edge.i]: Spacing here

**Review by permcody — CHANGES_REQUESTED** [review]
All the code looks good. You managed to propagate a few formatting issues in most of your input files though.

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_inner_tip.i (hunk 2 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_inner_tip.i]: done

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_lode_zero.i (hunk 3 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_lode_zero.i]: done

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_native.i (hunk 5 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_native.i]: done

File: modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_outer_tip.i (hunk 7 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/tests/capped_drucker_prager/small_deform2_outer_tip.i]: done

**Review by permcody — APPROVED** [review]

**Pull request merged**